File: wpilibc/src/test/native/cpp/simulation/AnalogGyroSimTest.cpp
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

#include "frc/simulation/AnalogGyroSim.h"  // NOLINT(build/include_order)

#include <memory>

#include <gtest/gtest.h>
#include <hal/HAL.h>

#include "callback_helpers/TestCallbackHelpers.h"
#include "frc/AnalogGyro.h"
#include "frc/AnalogInput.h"

namespace frc::sim {

TEST(AnalogGyroSimTest, InitializeGyro) {
  HAL_Initialize(500, 0);
  AnalogGyroSim sim{0};
  EXPECT_FALSE(sim.GetInitialized());

  BooleanCallback initializedCallback;

  auto cb =
      sim.RegisterInitializedCallback(initializedCallback.GetCallback(), false);
  AnalogGyro gyro(0);
  EXPECT_TRUE(sim.GetInitialized());
  EXPECT_TRUE(initializedCallback.WasTriggered());
  EXPECT_TRUE(initializedCallback.GetLastValue());
}

TEST(AnalogGyroSimTest, SetAngle) {
  HAL_Initialize(500, 0);

  AnalogGyroSim sim{0};
  DoubleCallback callback;

  AnalogInput ai(0);
  AnalogGyro gyro(&ai);
  auto cb = sim.RegisterAngleCallback(callback.GetCallback(), false);
  EXPECT_EQ(0, gyro.GetAngle());

  constexpr double kTestAngle = 35.04;
  sim.SetAngle(kTestAngle);
  EXPECT_EQ(kTestAngle, sim.GetAngle());
  EXPECT_EQ(kTestAngle, gyro.GetAngle());
  EXPECT_EQ(-kTestAngle, gyro.GetRotation2d().Degrees().value());
  EXPECT_TRUE(callback.WasTriggered());
  EXPECT_EQ(kTestAngle, callback.GetLastValue());
}

TEST(AnalogGyroSimTest, SetRate) {
  HAL_Initialize(500, 0);

  AnalogGyroSim sim{0};
  DoubleCallback callback;

  std::shared_ptr<AnalogInput> ai(new AnalogInput(0));
  AnalogGyro gyro(ai);
  auto cb = sim.RegisterRateCallback(callback.GetCallback(), false);
  EXPECT_EQ(0, gyro.GetRate());

  constexpr double kTestRate = -19.1;
  sim.SetRate(kTestRate);
  EXPECT_EQ(kTestRate, sim.GetRate());
  EXPECT_EQ(kTestRate, gyro.GetRate());

  EXPECT_TRUE(callback.WasTriggered());
  EXPECT_EQ(kTestRate, callback.GetLastValue());
}

TEST(AnalogGyroSimTest, Reset) {
  HAL_Initialize(500, 0);

  AnalogInput ai{0};
  AnalogGyro gyro(&ai);
  AnalogGyroSim sim(gyro);
  sim.SetAngle(12.34);
  sim.SetRate(43.21);

  sim.ResetData();
  EXPECT_EQ(0, sim.GetAngle());
  EXPECT_EQ(0, sim.GetRate());
  EXPECT_EQ(0, gyro.GetAngle());
  EXPECT_EQ(0, gyro.GetRate());
}
}  // namespace frc::sim

File: wpilibj/src/test/java/edu/wpi/first/wpilibj/simulation/AnalogGyroSimTest.java
// Copyright (c) FIRST and other WPILib contributors.
// Open Source Software; you can modify and/or share it under the terms of
// the WPILib BSD license file in the root directory of this project.

package edu.wpi.first.wpilibj.simulation;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertFalse;
import static org.junit.jupiter.api.Assertions.assertTrue;

import edu.wpi.first.hal.HAL;
import edu.wpi.first.wpilibj.AnalogGyro;
import edu.wpi.first.wpilibj.AnalogInput;
import edu.wpi.first.wpilibj.simulation.testutils.BooleanCallback;
import edu.wpi.first.wpilibj.simulation.testutils.DoubleCallback;
import org.junit.jupiter.api.Test;

class AnalogGyroSimTest {
  @Test
  void testInitialization() {
    HAL.initialize(500, 0);
    AnalogGyroSim sim = new AnalogGyroSim(0);
    assertFalse(sim.getInitialized());

    BooleanCallback initializedCallback = new BooleanCallback();

    try (CallbackStore cb = sim.registerInitializedCallback(initializedCallback, false);
        AnalogInput ai = new AnalogInput(0);
        AnalogGyro gyro = new AnalogGyro(ai)) {
      assertTrue(sim.getInitialized());
      assertTrue(initializedCallback.wasTriggered());
      assertTrue(initializedCallback.getSetValue());
    }
  }

  @Test
  void testSetAngle() {
    HAL.initialize(500, 0);

    AnalogGyroSim sim = new AnalogGyroSim(0);
    DoubleCallback callback = new DoubleCallback();

    try (AnalogInput ai = new AnalogInput(0);
        AnalogGyro gyro = new AnalogGyro(ai);
        CallbackStore cb = sim.registerAngleCallback(callback, false)) {
      assertEquals(0, gyro.getAngle());

      final double TEST_ANGLE = 35.04;
      sim.setAngle(TEST_ANGLE);
      assertEquals(TEST_ANGLE, sim.getAngle());
      assertEquals(TEST_ANGLE, gyro.getAngle());
      assertEquals(-TEST_ANGLE, gyro.getRotation2d().getDegrees());
      assertTrue(callback.wasTriggered());
      assertEquals(TEST_ANGLE, callback.getSetValue());
    }
  }

  @Test
  void testSetRate() {
    HAL.initialize(500, 0);

    AnalogGyroSim sim = new AnalogGyroSim(0);
    DoubleCallback callback = new DoubleCallback();

    try (AnalogInput ai = new AnalogInput(0);
        AnalogGyro gyro = new AnalogGyro(ai);
        CallbackStore cb = sim.registerRateCallback(callback, false)) {
      assertEquals(0, gyro.getRate());

      final double TEST_RATE = -19.1;
      sim.setRate(TEST_RATE);
      assertEquals(TEST_RATE, sim.getRate());
      assertEquals(TEST_RATE, gyro.getRate());

      assertTrue(callback.wasTriggered());
      assertEquals(TEST_RATE, callback.getSetValue());
    }
  }

  @Test
  void testReset() {
    HAL.initialize(500, 0);

    try (AnalogInput ai = new AnalogInput(0);
        AnalogGyro gyro = new AnalogGyro(ai)) {
      AnalogGyroSim sim = new AnalogGyroSim(gyro);
      sim.setAngle(12.34);
      sim.setRate(43.21);

      sim.resetData();
      assertEquals(0, sim.getAngle());
      assertEquals(0, sim.getRate());
      assertEquals(0, gyro.getAngle());
      assertEquals(0, gyro.getRate());
    }
  }
}
